File: amnmobile/lib/models/turno.dart
import 'package:json_annotation/json_annotation.dart';

part 'turno.g.dart';

@JsonSerializable()
class Turno {
  final int? id;
  final DateTime fecha;
  final String tipoTurno; // mañana/tarde/noche
  final String operador;
  final String supervisor;
  final String lineaProduccion;
  final String producto;
  final String lote;
  final int cantidadProducida;
  final int cantidadRechazada;
  final String? firmaDigital;
  final bool sincronizado;
  final double? latitud;
  final double? longitud;

  Turno({
    this.id,
    required this.fecha,
    required this.tipoTurno,
    required this.operador,
    required this.supervisor,
    required this.lineaProduccion,
    required this.producto,
    required this.lote,
    required this.cantidadProducida,
    required this.cantidadRechazada,
    this.firmaDigital,
    this.sincronizado = false,
    this.latitud,
    this.longitud,
  });

  factory Turno.fromJson(Map<String, dynamic> json) => Turno(
        id: json['id'] as int?,
        fecha: DateTime.parse(json['fecha'] as String),
        tipoTurno: json['tipoTurno'] as String,
        operador: json['operador'] as String,
        supervisor: json['supervisor'] as String,
        lineaProduccion: json['lineaProduccion'] as String,
        producto: json['producto'] as String,
        lote: json['lote'] as String,
        cantidadProducida: json['cantidadProducida'] as int,
        cantidadRechazada: json['cantidadRechazada'] as int,
        firmaDigital: json['firmaDigital'] as String?,
        sincronizado: (json['sincronizado'] is int)
            ? json['sincronizado'] == 1
            : (json['sincronizado'] as bool? ?? false),
        latitud: (json['latitud'] as num?)?.toDouble(),
        longitud: (json['longitud'] as num?)?.toDouble(),
      );
  Map<String, dynamic> toJson() => _$TurnoToJson(this);

  Turno copyWith({
    int? id,
    DateTime? fecha,
    String? tipoTurno,
    String? operador,
    String? supervisor,
    String? lineaProduccion,
    String? producto,
    String? lote,
    int? cantidadProducida,
    int? cantidadRechazada,
    String? firmaDigital,
    bool? sincronizado,
    double? latitud,
    double? longitud,
  }) {
    return Turno(
      id: id ?? this.id,
      fecha: fecha ?? this.fecha,
      tipoTurno: tipoTurno ?? this.tipoTurno,
      operador: operador ?? this.operador,
      supervisor: supervisor ?? this.supervisor,
      lineaProduccion: lineaProduccion ?? this.lineaProduccion,
      producto: producto ?? this.producto,
      lote: lote ?? this.lote,
      cantidadProducida: cantidadProducida ?? this.cantidadProducida,
      cantidadRechazada: cantidadRechazada ?? this.cantidadRechazada,
      firmaDigital: firmaDigital ?? this.firmaDigital,
      sincronizado: sincronizado ?? this.sincronizado,
      latitud: latitud ?? this.latitud,
      longitud: longitud ?? this.longitud,
    );
  }
} 

File: amnappBackend/models/Turno.js
const mongoose = require('mongoose');

const turnoSchema = new mongoose.Schema({
  local_id: {
    type: String,
    required: true
  },
  fecha: {
    type: Date,
    required: true
  },
  tipo: {
    type: String,
    required: true,
    enum: ['mañana', 'tarde', 'noche']
  },
  operador: {
    type: String,
    required: true
  },
  supervisor: {
    type: String,
    required: true
  },
  linea: {
    type: String,
    required: true
  },
  producto: {
    type: String,
    required: true
  },
  lote: {
    type: String,
    required: true
  },
  cantidad_producida: {
    type: Number,
    required: true,
    min: 0
  },
  cantidad_rechazada: {
    type: Number,
    required: true,
    min: 0
  },
  observaciones: String,
  firma_digital: {
    type: String,
    required: true
  },
  sincronizado: {
    type: Boolean,
    default: false
  },
  fecha_sincronizacion: Date
}, {
  timestamps: true
});

// Índices
turnoSchema.index({ local_id: 1 }, { unique: true });
turnoSchema.index({ fecha: 1, tipo: 1, linea: 1 });
turnoSchema.index({ sincronizado: 1 });

module.exports = mongoose.model('Turno', turnoSchema); 